File: Xamarin/SampleXamarin.Android/AzureSpatialAnchorsManager.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using Google.AR.Core;
using Java.Util.Concurrent;
using Microsoft.Azure.SpatialAnchors;
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;

namespace SampleXamarin
{
    internal class AzureSpatialAnchorsManager
    {
        private readonly CloudSpatialAnchorSession spatialAnchorsSession;

        private TrackingState lastTrackingState = TrackingState.Stopped;

        private TrackingFailureReason lastTrackingFailureReason = TrackingFailureReason.None;

        public bool CanCreateAnchor => this.CreateScanningProgressValue >= 1;

        public float CreateScanningProgressValue { get; set; } = 0;

        public bool LocatingAnchors => this.spatialAnchorsSession.ActiveWatchers.Count > 0;

        public bool IsArTracking => this.lastTrackingState == TrackingState.Tracking;

        public bool Running { get; private set; }

        public event EventHandler<AnchorLocatedEvent> OnAnchorLocated;

        public event EventHandler<LocateAnchorsCompletedEvent> OnLocateAnchorsCompleted;

        public event EventHandler<LogDebugEventArgs> OnLogDebug;

        public event EventHandler<SessionErrorEvent> OnSessionError;

        public event EventHandler<SessionUpdatedEvent> OnSessionUpdated;

        static AzureSpatialAnchorsManager()
        {
            CloudServices.Initialize(Android.App.Application.Context);
        }

        public AzureSpatialAnchorsManager(Session arCoreSession)
        {
            this.spatialAnchorsSession = new CloudSpatialAnchorSession();
            this.spatialAnchorsSession.Configuration.AccountKey = AccountDetails.SpatialAnchorsAccountKey;
            this.spatialAnchorsSession.Configuration.AccountId = AccountDetails.SpatialAnchorsAccountId;
            this.spatialAnchorsSession.Session = arCoreSession;
            this.spatialAnchorsSession.LogDebug += this.SpatialCloudSession_LogDebug;
            this.spatialAnchorsSession.Error += this.SpatialAnchorsSession_Error;
            this.spatialAnchorsSession.AnchorLocated += this.SpatialAnchorsSession_AnchorLocated;
            this.spatialAnchorsSession.LocateAnchorsCompleted += this.SpatialAnchorsSession_LocateAnchorsCompleted;
            this.spatialAnchorsSession.SessionUpdated += this.SpatialAnchorsSession_SessionUpdated;
        }

        public CloudSpatialAnchorWatcher StartLocating(AnchorLocateCriteria locateCriteria)
        {
            // Only 1 active watcher at a time is permitted.
            this.StopLocating();

            return this.spatialAnchorsSession.CreateWatcher(locateCriteria);
        }

        public Task DeleteAnchorAsync(CloudSpatialAnchor anchor)
        {
            return this.spatialAnchorsSession.DeleteAnchorAsync(anchor).GetAsync();
        }

        public void ResetSession(bool resumeIfRunning = false)
        {
            bool running = this.Running;

            this.StopLocating();
            this.StopSession();
            this.spatialAnchorsSession.Reset();

            if (resumeIfRunning && running)
            {
                this.StartSession();
            }
        }

        public async Task<CloudSpatialAnchor> CreateAnchorAsync(CloudSpatialAnchor newCloudAnchor)
        {
            if (newCloudAnchor == null)
            {
                throw new ArgumentNullException(nameof(newCloudAnchor));
            }

            if (newCloudAnchor.LocalAnchor == null || !string.IsNullOrEmpty(newCloudAnchor.Identifier))
            {
                throw new ArgumentException("The specified cloud anchor cannot be saved.", nameof(newCloudAnchor));
            }

            if (!this.CanCreateAnchor)
            {
                throw new ArgumentException("Not ready to create. Need more data.");
            }

            try
            {
                await this.spatialAnchorsSession.CreateAnchorAsync(newCloudAnchor).GetAsync();
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex.Message);
            }

            return newCloudAnchor;
        }

        public void StartSession()
        {
            if (!this.Running)
            {
                this.spatialAnchorsSession.Start();
                this.Running = true;
            }
        }

        public void StopLocating()
        {
            CloudSpatialAnchorWatcher watcher = this.spatialAnchorsSession.ActiveWatchers.FirstOrDefault();

            // Only 1 active watcher at a time is permitted.
            watcher?.Stop();
            watcher?.Dispose();
        }

        public void StopSession()
        {
            this.StopLocating();
            if (this.Running)
            {
                this.spatialAnchorsSession.Stop();
                this.Running = false;
            }
        }

        public void Update(Frame frame)
        {
            if (frame.Camera.TrackingState != this.lastTrackingState
                || frame.Camera.TrackingFailureReason != this.lastTrackingFailureReason)
            {
                this.lastTrackingState = frame.Camera.TrackingState;
                this.lastTrackingFailureReason = frame.Camera.TrackingFailureReason;
                Debug.WriteLine($"Tracker state changed: {this.lastTrackingState}, {this.lastTrackingFailureReason}.");
            }

            Task.Run(() => this.spatialAnchorsSession.ProcessFrame(frame));
        }

        private void SpatialAnchorsSession_AnchorLocated(object sender, AnchorLocatedEventArgs e)
        {
            this.OnAnchorLocated?.Invoke(sender, e.Args);
        }

        private void SpatialAnchorsSession_Error(object sender, SessionErrorEventArgs e)
        {
            SessionErrorEvent eventArgs = e?.Args;

            if (eventArgs == null)
            {
                Debug.WriteLine("Azure Spatial Anchors reported an unspecified error.");
                return;
            }

            string message = $"{eventArgs.ErrorCode}: {eventArgs.ErrorMessage}";
            Debug.WriteLine(message);

            this.OnSessionError?.Invoke(sender, eventArgs);
        }

        private void SpatialAnchorsSession_LocateAnchorsCompleted(object sender, LocateAnchorsCompletedEventArgs e)
        {
            this.OnLocateAnchorsCompleted?.Invoke(sender, e.Args);
        }

        private void SpatialAnchorsSession_SessionUpdated(object sender, SessionUpdatedEventArgs e)
        {
            float createScanProgress = Math.Min(e.Args.Status.RecommendedForCreateProgress, 1);

            Debug.WriteLine($"Create scan progress: {createScanProgress:0%}");

            this.CreateScanningProgressValue = createScanProgress;

            this.OnSessionUpdated?.Invoke(sender, e.Args);
        }

        private void SpatialCloudSession_LogDebug(object sender, LogDebugEventArgs e)
        {
            if (string.IsNullOrWhiteSpace(e.Args.Message))
            {
                return;
            }

            Debug.WriteLine(e.Args.Message);

            this.OnLogDebug?.Invoke(sender, e);
        }
    }
}

File: Android/Java/app/src/main/java/com/microsoft/sampleandroid/AzureSpatialAnchorsManager.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
package com.microsoft.sampleandroid;

import android.util.Log;

import com.google.ar.core.Frame;
import com.google.ar.core.Session;
import com.microsoft.azure.spatialanchors.AnchorLocateCriteria;
import com.microsoft.azure.spatialanchors.AnchorLocatedListener;
import com.microsoft.azure.spatialanchors.CloudSpatialAnchor;
import com.microsoft.azure.spatialanchors.CloudSpatialAnchorSession;
import com.microsoft.azure.spatialanchors.CloudSpatialAnchorWatcher;
import com.microsoft.azure.spatialanchors.PlatformLocationProvider;
import com.microsoft.azure.spatialanchors.LocateAnchorsCompletedListener;
import com.microsoft.azure.spatialanchors.OnLogDebugEvent;
import com.microsoft.azure.spatialanchors.SessionErrorEvent;
import com.microsoft.azure.spatialanchors.SessionLogLevel;
import com.microsoft.azure.spatialanchors.SessionUpdatedListener;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

class AzureSpatialAnchorsManager {
    // Set this string to the account ID provided for the Azure Spatial Service resource.
    public static final String SpatialAnchorsAccountId = "Set me";

    // Set this string to the account key provided for the Azure Spatial Service resource.
    public static final String SpatialAnchorsAccountKey = "Set me";

    // Log message tag
    private static final String TAG = "ASACloud";

    private final ExecutorService executorService = Executors.newFixedThreadPool(2);

    private boolean running = false;

    private final CloudSpatialAnchorSession spatialAnchorsSession;

    public AzureSpatialAnchorsManager(Session arCoreSession) {
        if (arCoreSession == null) {
            throw new IllegalArgumentException("The arCoreSession may not be null.");
        }

        spatialAnchorsSession = new CloudSpatialAnchorSession();
        spatialAnchorsSession.getConfiguration().setAccountId(SpatialAnchorsAccountId);
        spatialAnchorsSession.getConfiguration().setAccountKey(SpatialAnchorsAccountKey);
        spatialAnchorsSession.setSession(arCoreSession);
        spatialAnchorsSession.setLogLevel(SessionLogLevel.All);

        spatialAnchorsSession.addOnLogDebugListener(this::onLogDebugListener);
        spatialAnchorsSession.addErrorListener(this::onErrorListener);
    }

    //region Listener Handling

    public void addSessionUpdatedListener(SessionUpdatedListener listener) {
        this.spatialAnchorsSession.addSessionUpdatedListener(listener);
    }

    public void removeSessionUpdatedListener(SessionUpdatedListener listener) {
        this.spatialAnchorsSession.removeSessionUpdatedListener(listener);
    }

    public void addAnchorLocatedListener(AnchorLocatedListener listener) {
        this.spatialAnchorsSession.addAnchorLocatedListener(listener);
    }

    public void removeAnchorLocatedListener(AnchorLocatedListener listener) {
        this.spatialAnchorsSession.removeAnchorLocatedListener(listener);
    }

    public void addLocateAnchorsCompletedListener(LocateAnchorsCompletedListener listener) {
        this.spatialAnchorsSession.addLocateAnchorsCompletedListener(listener);
    }

    public void removeLocateAnchorsCompletedListener(LocateAnchorsCompletedListener listener) {
        this.spatialAnchorsSession.removeLocateAnchorsCompletedListener(listener);
    }

    //endregion

    public void setLocationProvider(PlatformLocationProvider locationProvider) {
        spatialAnchorsSession.setLocationProvider(locationProvider);
    }

    public CompletableFuture<CloudSpatialAnchor> createAnchorAsync(CloudSpatialAnchor anchor) {
        //noinspection unchecked,unchecked
        return this.toEmptyCompletableFuture(spatialAnchorsSession.createAnchorAsync(anchor))
                .thenApply((ignore) -> anchor);
    }

    public CompletableFuture deleteAnchorAsync(CloudSpatialAnchor anchor) {
        return this.toEmptyCompletableFuture(spatialAnchorsSession.deleteAnchorAsync(anchor));
    }

    public boolean isRunning() {
        return this.running;
    }

    public void reset() {
        stopLocating();
        spatialAnchorsSession.reset();
    }

    public void start() {
        spatialAnchorsSession.start();
        this.running = true;
    }

    public CloudSpatialAnchorWatcher startLocating(AnchorLocateCriteria criteria) {
        // Only 1 active watcher at a time is permitted.
        stopLocating();

        return spatialAnchorsSession.createWatcher(criteria);
    }

    public boolean isLocating() {
        return !spatialAnchorsSession.getActiveWatchers().isEmpty();
    }

    public void stopLocating() {
        List<CloudSpatialAnchorWatcher> watchers = spatialAnchorsSession.getActiveWatchers();

        if (watchers.isEmpty()) {
            return;
        }

        // Only 1 watcher is at a time is currently permitted.
        CloudSpatialAnchorWatcher watcher = watchers.get(0);

        watcher.stop();
    }

    public void stop() {
        spatialAnchorsSession.stop();
        stopLocating();
        this.running = false;
    }

    public void update(Frame frame) {
        spatialAnchorsSession.processFrame(frame);
    }

    private <T> CompletableFuture<T> toCompletableFuture(Future<T> future) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                return future.get();
            } catch (InterruptedException|ExecutionException e) {
                e.printStackTrace();
                throw new RuntimeException(e);
            }
        }, executorService);
    }

    private CompletableFuture toEmptyCompletableFuture(Future future) {
        return CompletableFuture.runAsync(() -> {
            try {
                future.get();
            } catch (InterruptedException|ExecutionException e) {
                e.printStackTrace();
                throw new RuntimeException(e);
            }
        }, executorService);
    }

    private void onErrorListener(SessionErrorEvent event) {
        Log.e(TAG, event.getErrorMessage());
    }

    private void onLogDebugListener(OnLogDebugEvent args) {
        Log.d(TAG, args.getMessage());
    }
}
